### PR TITLE
[REVIEW] Fixed memory leaks in urldecode() and insert()

### LINKED ISSUE #408: [BUG] NVStrings insert method create instance with incorrect device_memory size
This size is used for reporting only for now.
The accuracy of this value may be critical for the IPC transfer fix.
https://github.com/rapidsai/cudf/issues/2469
Also, future optimizations can use this value more efficiently copy strings to CPU (e.g. in the csv-writer).

### LINKED ISSUE #2451: [BUG] Extra memory footprint by nvstrings.url_decode causes dask to OOM
**Describe the bug**
The following is a minimal reproducible sample of a subset of the actual dataset(private) hence not posting the dataset and instead posting a near replica of the data.

**Steps/Code to reproduce bug**

The GPU memory usage for a string replace of a column with 10038326 records is at `3429MiB`, while if we use `url_decode` I find that there is extra memory usage(`3733MiB`) which is potentially causing dask to OOM with very large datasets(1075366479 records). I have verified the removing the `url_decode` call from the workflow is avoiding the OOM.

`url_decode`
```python
import cudf
df = cudf.DataFrame({'path':['esjfdlk jslkdfj l ldsjk %2Fslk sdkshk ajkhdk s hdksah aksjhdkjs dkjkhjhsad djflkdjlk7%3D%20%3A%2B%24dgff djlfk %3B%23%2C ksdjhfkjdshfkjdshf22%25%21%5B%5D%3F%3E%7C'] * 10038326})
y = df.path.str.url_decode()
```

GPU usage at this point:
```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0    346429      C   /opt/conda/envs/rapids/bin/python            3733MiB|
+-----------------------------------------------------------------------------+
```

After restarting kernel 

String replace:
```python
import cudf
df = cudf.DataFrame({'path':['esjfdlk jslkdfj l ldsjk %2Fslk sdkshk ajkhdk s hdksah aksjhdkjs dkjkhjhsad djflkdjlk7%3D%20%3A%2B%24dgff djlfk %3B%23%2C ksdjhfkjdshfkjdshf22%25%21%5B%5D%3F%3E%7C'] * 10038326})
y = df.path.str.replace('%2F','/').str.replace('%27',"\'").str.replace('%3D','=').str.replace('%20',' ').str.replace('%3A',':').str.replace('%2B','+').str.replace('%24','$').str.replace('%3B',';').str.replace('%23','#').str.replace('%2C',',').str.replace('%2A','*').str.replace('%40','@').str.replace('%28','(').str.replace('%29',')').str.replace('%5E','^').str.replace('%5C','\\').str.replace('%22','"').str.replace('%25','%').str.replace('%21','!').str.replace('%5B','[').str.replace('%5D',']').str.replace('%3F','?').str.replace('%3E','>').str.replace('%7C','|')
```

GPU usage at this point:

```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0    347047      C   /opt/conda/envs/rapids/bin/python            3429MiB|
+-----------------------------------------------------------------------------+
```

Verified the parity between two outputs as well:
```python
import cudf
df = cudf.DataFrame({'path':['esjfdlk jslkdfj l ldsjk %2Fslk sdkshk ajkhdk s hdksah aksjhdkjs dkjkhjhsad djflkdjlk7%3D%20%3A%2B%24dgff djlfk %3B%23%2C ksdjhfkjdshfkjdshf22%25%21%5B%5D%3F%3E%7C'] * 10038326})
y1 = df.path.str.replace('%2F','/').str.replace('%27',"\'").str.replace('%3D','=').str.replace('%20',' ').str.replace('%3A',':').str.replace('%2B','+').str.replace('%24','$').str.replace('%3B',';').str.replace('%23','#').str.replace('%2C',',').str.replace('%2A','*').str.replace('%40','@').str.replace('%28','(').str.replace('%29',')').str.replace('%5E','^').str.replace('%5C','\\').str.replace('%22','"').str.replace('%25','%').str.replace('%21','!').str.replace('%5B','[').str.replace('%5D',']').str.replace('%3F','?').str.replace('%3E','>').str.replace('%7C','|')
y2 = df.path.str.url_decode()
print((y1 == y2).value_counts())
True    10038326
dtype: int64
```

**Expected behavior**
Same memory foot-print to avoid OOM when performing a scaled run with dask.

**Environment overview (please complete the following information)**
 - Environment location: Docker
 - Method of cuDF install: from source

**Environment details**
 `cudf/print_env.sh`: [env_aug1.txt](https://github.com/rapidsai/cudf/files/3458692/env_aug1.txt)

**Additional context**
Provided in the above along with steps.

Logging this issue in cudf as custrings is about to be merged into cudf. Hence logged this issue here. 

cc: @davidwendt @beckernick @kkraus14

### PR DESCRIPTION
**Pull request by davidwendt**
Fixes https://github.com/rapidsai/cudf/issues/2451
Closes #408

### CODE REVIEW

**Review by jrhemstad — COMMENTED** [review]
It's difficult for me to meaningfully review this code because I don't know what the operations are supposed to be doing and there is no documentation to tell me what they are doing.

**Comment by davidwendt** [comment]
Here is an example of `urldecode` in action: https://github.com/rapidsai/custrings/blob/b1913fa587fadc1ebdbcbcdfd03e984d8af00126/cpp/tests/test_url.cpp#L27
The issue as described in https://github.com/rapidsai/cudf/issues/2451 is that it is over-allocating memory.
Here is an example of `insert`: https://github.com/rapidsai/custrings/blob/b1913fa587fadc1ebdbcbcdfd03e984d8af00126/cpp/tests/test_modify.cu#L96
This one was also computing the memory size incorrectly (too large).

I'm happy to talk through any of these.

**Comment by jrhemstad** [comment]
Can you add documentation to the `url_encoder` and `url_decoder` function objects?

**Comment by davidwendt** [comment]
> Can you add documentation to the `url_encoder` and `url_decoder` function objects?

Generally no one calls these C++ methods right now except `nvstrings.py` and so they are documented here: https://rapidsai.github.io/projects/nvstrings/en/0.9.0/api.html#nvstrings.nvstrings.url_decode

Do you want me to add examples to the C++ doxygen comments in the NVStrings.h for these 2 methods?

**Comment by jrhemstad** [comment]
I was suggesting documenting the internal `struct`s used to implement the higher level API, i.e., documenting internal-only functions. 

I always do this in libcudf and ask others to do the same.

File: cpp/src/strings/urlencode.cu [hunk 1]
```diff
@@ -30,6 +30,17 @@
 #include "../util.h"
 
 //
+// This is the functor for the url_encode() method below.
```
Comment (jrhemstad) [inline cpp/src/strings/urlencode.cu:33]: This is great! Thank you.

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**